src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java

```diff
@@ -30,143 +38,272 @@
+        Set<String> virtualFieldNames = tableMapping.getAllMappingsPerField().keySet();
+
+        updateExpression = convertFieldNameLiteralsToExpressionNames(updateExpression, request, virtualFieldNames);
+        validateNotUpdatingTablePrimaryKeyFields(updateExpression, request);
+
+        if (conditionExpression != null) {
+            conditionExpression = convertFieldNameLiteralsToExpressionNames(conditionExpression, request,
+                virtualFieldNames);
+            conditionExpression = makePlaceholdersDistinct(updateExpression, conditionExpression, request);
+            request.setFilterExpression(conditionExpression);
+            applyToFilterExpression(request, false);
+        }
+
+        for (Entry<String, List<FieldMapping>> entry : tableMapping.getAllMappingsPerField().entrySet()) {
+            updateExpression = mapFieldInUpdateExpression(updateExpression, request, entry.getKey(), entry.getValue());
+        }
+        request.setPrimaryExpression(updateExpression);
     }
 
     @Override
-    public void apply(RequestWrapper request) {
-        apply(request, null);
+    public void applyToKeyCondition(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex) {
+        Collection<FieldMapping> fieldMappings = virtualSecondaryIndex == null
+            ? tableMapping.getTablePrimaryKeyFieldMappings()
+            : tableMapping.getIndexPrimaryKeyFieldMappings(virtualSecondaryIndex);
+
+        String expression = request.getPrimaryExpression();
+        expression = mapFieldsInConditionExpression(expression, request, fieldMappings);
+        request.setPrimaryExpression(expression);
     }
 
     @Override
-    public void apply(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex) {
-        Collection<FieldMapping> fieldMappings;
-        if (virtualSecondaryIndex == null) {
-            fieldMappings = tableMapping.getTablePrimaryKeyFieldMappings();
+    public void applyToFilterExpression(RequestWrapper request, boolean isPrimaryExpression) {
+        String expression = isPrimaryExpression ? request.getPrimaryExpression() : request.getFilterExpression();
+
+        // for conditional / filter expressions, for each indexed field, we only need to use one of its field mappings,
+        // since we're supporting only equals expressions
+        List<FieldMapping> fieldMappings = new ArrayList<>(tableMapping.getAllMappingsPerField().size());
+        tableMapping.getAllMappingsPerField().forEach((virtualField, mappings) -> fieldMappings.add(mappings.get(0)));
+
+        expression = mapFieldsInConditionExpression(expression, request, fieldMappings);
+
+        if (isPrimaryExpression) {
+            request.setPrimaryExpression(expression);
         } else {
-            fieldMappings = tableMapping.getIndexPrimaryKeyFieldMappings(virtualSecondaryIndex);
+            request.setFilterExpression(expression);
         }
+    }
+
```

Comment:
The method below is used by applyToFilterExpression() and by applyToKeyCondition() -- will add comment in code